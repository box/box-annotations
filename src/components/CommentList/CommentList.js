// @flow
import React from 'react';

import Comment from '../Comment';

import './CommentList.scss';

const CLASS_COMMENT_LIST = 'ba-comment-list';
const CLASS_COMMENT_LIST_ITEM = 'ba-comment-list-item';

type Props = {
    comments: Comments,
    onDelete: Function
};

class CommentList extends React.Component<Props> {
    commentsContainer: null | HTMLElement;

    componentDidMount() {
        this.scrollToBottom();
    }

    componentDidUpdate() {
        this.scrollToBottom();
    }

    /**
     * Scrolls the container to the bottom
     * @return {void}
     */
    scrollToBottom = () => {
        if (this.commentsContainer) {
            this.commentsContainer.scrollTop = this.commentsContainer.scrollHeight;
        }
    };

    render() {
        const { comments, onDelete } = this.props;
        return (
            <ul
                className={CLASS_COMMENT_LIST}
                ref={(ref) => {
                    this.commentsContainer = ref;
                }}
            >
                {comments.map(({ id, ...rest }) => (
                    <li className={CLASS_COMMENT_LIST_ITEM} key={`annotation_${id}`}>
                        <Comment id={id} onDelete={onDelete} {...rest} />
                    </li>
                ))}
            </ul>
        );
    }
}

export default CommentList;
