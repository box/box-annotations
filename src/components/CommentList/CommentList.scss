.ba {
    @import "../../variables.scss";

    .ba-comment-list {
        width: 100%;
        max-height: 282px;
        overflow-y: auto;
        user-select: text;
    }

    .ba-comment-list-item {
        @import "~box-react-ui/lib/styles/common/inline-notifications";

        margin-bottom: 15px; // Matches spacing of activity feed items in UI Elements
        list-style: none;
    }

    // Spaces comment elements from each other and the dividing border
    // Adds extra spacing above consecutive annotations as well
    .ba-comment-list-item + .ba-comment-list-item {
        padding-top: 15px;
        border-top: 1px solid $seesee;
    }

    .bcs-comment,
    .bcs-comment-user-name ~ .bcs-comment-created-at {
        margin: 0;
    }

    .bcs-comment-content {
        text-align: left;
    }

    // MOBILE CSS
    @media #{$mobile}, #{$tablet} {
        .ba-comment-list {
            flex-grow: 1;
            width: unset;
            max-height: unset;
            padding: 15px;
            background: white;
        }

        .ba-comment-list-item {
            margin-bottom: 30px;
            padding-top: 15px;
        }

        .ba-comment-list-item + .ba-comment-list-item {
            padding-top: 30px;
        }
    }
}
