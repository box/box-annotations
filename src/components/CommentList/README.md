### Description
Component to represent a list of comments

### Examples
```jsx
const CommentListExample = require('examples').CommentListExample;

<CommentListExample />
```
