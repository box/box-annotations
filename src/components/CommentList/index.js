// @flow
export { default } from './CommentList';
