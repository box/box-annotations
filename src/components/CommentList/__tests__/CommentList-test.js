import * as React from 'react';
import { shallow } from 'enzyme';

import CommentList from '../CommentList';

const TIME_STRING_SEPT_27_2017 = '2017-09-27T10:40:41-07:00';

const USER = {
    type: 'user',
    id: '789ghi',
    name: 'Mia Thermopolis',
    email: 'princess@genovia.gov'
};

const comments = [
    {
        id: '123',
        createdAt: TIME_STRING_SEPT_27_2017,
        createdBy: USER,
        message: 'test',
        permissions: {}
    },
    {
        id: '456',
        createdAt: TIME_STRING_SEPT_27_2017,
        createdBy: USER,
        message: 'test',
        permissions: {}
    }
];

const onDelete = jest.fn();

describe('components/CommentList', () => {
    test('should correctly render a list of comments', () => {
        const wrapper = shallow(<CommentList comments={comments} onDelete={onDelete} />);
        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.ba-comment-list-item').length).toEqual(2);
    });

    test('should scroll to the bottom on componentDidUpdate()', () => {
        const wrapper = shallow(<CommentList comments={comments} onDelete={onDelete} />);
        const instance = wrapper.instance();
        instance.scrollToBottom = jest.fn();

        wrapper.setProps({ comments: [...comments, ...comments] });
        expect(instance.scrollToBottom).toBeCalled();
    });
});
