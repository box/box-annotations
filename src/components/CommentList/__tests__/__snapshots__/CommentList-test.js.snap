// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/CommentList should correctly render a list of comments 1`] = `
<ul
  className="ba-comment-list"
>
  <li
    className="ba-comment-list-item"
    key="annotation_123"
  >
    <Comment
      createdAt="2017-09-27T10:40:41-07:00"
      createdBy={
        Object {
          "email": "princess@genovia.gov",
          "id": "789ghi",
          "name": "Mia Thermopolis",
          "type": "user",
        }
      }
      id="123"
      isPending={false}
      message="test"
      onDelete={[MockFunction]}
      permissions={Object {}}
    />
  </li>
  <li
    className="ba-comment-list-item"
    key="annotation_456"
  >
    <Comment
      createdAt="2017-09-27T10:40:41-07:00"
      createdBy={
        Object {
          "email": "princess@genovia.gov",
          "id": "789ghi",
          "name": "Mia Thermopolis",
          "type": "user",
        }
      }
      id="456"
      isPending={false}
      message="test"
      onDelete={[MockFunction]}
      permissions={Object {}}
    />
  </li>
</ul>
`;
